### PR TITLE
Fix handling of geostationary x/y units in CF conversion

### LINKED ISSUE #355: Handle X/Y in meters for CF conversion
#### Problem description

This was brought up on slack when working with Satpy's `cf` writer and then trying to read it back. The `cf` writer uses `AreaDefinition.crs.to_cf` [here](https://github.com/pytroll/satpy/blob/153d8685699f546e6a4f91d06a2327b8efed651d/satpy/writers/cf_writer.py#L154) to convert the CRS information to a grid_mapping variable. I'm not really sure where the x/y variables are handled. If a user then tries to read that NetCDF file back and get the area using `AreaDefinition.from_cf` they get the wrong units from the extent because `_convert_XY_CF_to_Proj` assumes that geostationary projections are in radians, but they can also be in microradians or meters. It also isn't clear to me why it sets the `units` to `None` when the comment above it says it is meters. Shouldn't it just be explicit and set it to meters?

https://github.com/pytroll/pyresample/blob/91b8ba79461af9ec15081e94b43ec4d82b4a3141/pyresample/utils/cf.py#L64-L77

Sorry, I don't have a code example right now. Note that at the time of writing the CF reader in Satpy doesn't read AreaDefinitions, but it should in the future.

### PR DESCRIPTION
**Pull request by zxdawn**
The geostationary projections can be radians, microradians or meters. We need to handle these in `_convert_XY_CF_to_Proj`.

 - [ ] Closes #355, [satpy issue 1688](https://github.com/pytroll/satpy/issues/1688)
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/main **/*py | flake8 --diff``

### CODE REVIEW

**Comment by zxdawn** [comment]
Find another special case of ABI L1 C13 data.

crs_dict:
```
{'proj': 'geos', 'sweep': 'x', 'lon_0': -75, 'h': 35786023, 'x_0': 0, 'y_0': 0, 'ellps': 'GRS80', 'units': 'm', 'no_defs': None, 'type': 'crs'}
```

axis_info:
```
{'first': 0.12821200489997864, 'last': 0.04426800459623337, 'spacing': 5.600000020263194e-05, 'nb': 1500, 'sign': -1.0, 'unit': None}
```

In this case, this issue still exists ... As the `units` in crs_dict leads to skipping the conversion of axis_info from radians to meters.

**Comment by djhoese** [comment]
Ok so the `if` check needs to be on `axis_info`? Or maybe any time the axis units are different than the CRS units?

**Comment by zxdawn** [comment]
But the `unit` is None in the ABI `axis_info`. Is it a bug of ABI L1 reader or `axis_info`?

**Comment by djhoese** [comment]
The `axis_info` is all coming from pyproj/PROJ, right? I think that is trying its hardest to follow CF conventions where in the original CF spec that supported the geostationary projection, it was in radians by default. So I don't think it is necessarily a bug anywhere. The CRS is defined to be in meters, the axis is defined to be `unit=None` (default to radians). I guess if anything maybe pyproj/PROJ should have the axis info actually list the default units.

**Comment by zxdawn** [comment]
Em ... Since the unit is missing, maybe we can just check whether `first` in `axis_info` is < 1?

**Comment by djhoese** [comment]
Seems error prone. You're basically saying if it is `<1` then it is probably in radians, right? Eh, I don't like that. Am I wrong about checking the units?

**Comment by zxdawn** [comment]
```
if crs_dict['proj'] == 'geos' and axis_info['first'] < 1:
```

This one works for me. I also don't like this method ... Then, is there any method to check whether the unit of `axis_info` is radians?

**Comment by djhoese** [comment]
Can't you do `axis_info.get('unit') or 'radians' == 'radians'` (not sure if parentheses are needed around the or chunk)?

**Comment by djhoese** [comment]
HTTP issue on the CI jobs. Restarted. Let's see if they pass. Is this ready for merge if this works (and I re-review it)?

File: pyresample/utils/cf.py [hunk 1]
```diff
@@ -64,15 +64,15 @@
...
+    if (crs_dict['proj'] == 'geos') and (axis_info.get('unit') or 'radians' == 'radians'):
         # for geostationary projection, the values stored as x/y in CF are not directly
         #  the x/y along the projection axes, but are rather the scanning angles from
         #  the satellite. We must multiply them by the height of the satellite.
         satellite_height = crs_dict['h']
         for k in ('first', 'last', 'spacing'):
             axis_info[k] *= satellite_height
         # the unit is now the default (meters)
-        axis_info['units'] = None
+        axis_info['units'] = 'meters'
```
Comment (djhoese) [inline pyresample/utils/cf.py:75]: Just thought of this, does pyproj/PROJ refer to this as `metre` or `m`? This is assuming that pyproj's CRS object is where we get the original version of this.

**Review by djhoese — CHANGES_REQUESTED** [review]
Small question inline, otherwise I think this just needs some tests. Thanks for fixing this and deciphering all of the units stuff.

File: pyresample/utils/cf.py (hunk 1 above)
Comment (zxdawn) [reply pyresample/utils/cf.py:75]: Thanks for the review. It should be `m`. I will fix it and add the test later.

**Comment by mraspaud** [comment]
Maybe this conversation it relevant here? https://github.com/cf-convention/cf-conventions/issues/230
TLDR; for cf > 1.7 we should support `projection_x_angular_coordinate` and `projection_y_angular_coordinate`

File: pyresample/utils/cf.py [hunk 2]
```diff
@@ -64,15 +64,15 @@
 def _convert_XY_CF_to_Proj(crs, axis_info):
     """Convert XY values from CF to PROJ convention. With CF =< 1.9 only affects geostrationary projection."""
     crs_dict = crs.to_dict()
-    if crs_dict['proj'] == 'geos':
+    if (crs_dict['proj'] == 'geos') and (axis_info.get('unit') or 'radians' == 'radians'):
```
Comment (djhoese) [inline pyresample/utils/cf.py:67]:
```suggestion
    axis_units = axis_info.get('unit') or 'radians'  # unit could be None
    if crs_dict['proj'] == 'geos' and axis_units == 'radians':
```

Going to try to make deepcode happy.

**Review by djhoese — APPROVED** [review]
Thanks again for this. This looks good. I made one small commit to make deepcode happy. It was getting confused by the `'radians' == 'radians'` part even though python interprets that correctly. I think once tests pass we can merge this.

**Comment by djhoese** [comment]
Wait...I just noticed there are no tests. Can you add some @zxdawn? Sorry, I got excited about this fix, but we should really have some tests to be sure this remains fixed in the future.

**Comment by djhoese** [comment]
Can you double check your tests? Coveralls and codecov are saying that `if` branch isn't actually being touched by any of the tests.

**Comment by djhoese** [comment]
Just ran them locally to make sure. Looks like the axis_info shows up as `m` so the `if` branch is never entered.

**Comment by zxdawn** [comment]
Oh, forgot to change the unit! Modified.

**Comment by djhoese** [comment]
Ok, thanks. Looks good to me, but it is late and I'd like to review this again when I'm not almost asleep.

File: pyresample/test/test_utils.py [hunk 3]
```diff
@@ -730,6 +778,33 @@ def validate_nh10km_cfinfo(cfinfo, variable='ice_conc', lat='lat', lon='lon'):
         _, cf_info = load_cf_area(cf_file)
         validate_nh10km_cfinfo(cf_info)
 
+    def test_load_cf_goes(self):
+        from pyresample.utils import load_cf_area
+
+        def validate_goes(adef, cfinfo, y='y', x='x'):
```
Comment (mraspaud) [inline pyresample/test/test_utils.py:784]: The `x` and `y` parameters don't seem to be used in this function?

File: pyresample/test/test_utils.py [hunk 4]
```diff
@@ -730,6 +778,33 @@ def validate_nh10km_cfinfo(cfinfo, variable='ice_conc', lat='lat', lon='lon'):
...
+            self.assertEqual(cf_info['y']['varname'], 'y')
+            self.assertEqual(cf_info['y']['last'], 1583173.6575)
+
+        # prepare xarray Dataset
+        cf_file = _prepare_cf_goes()
+
+        # load using a variable= that is a valid grid_mapping container
+        adef, cf_info = load_cf_area(cf_file)
+        validate_goes(adef, cf_info, y=None, x=None)
+
```
Comment (mraspaud) [inline pyresample/test/test_utils.py:799]:
```suggestion
```

File: pyresample/test/test_utils.py (hunk 3 above)
Comment (mraspaud) [reply pyresample/test/test_utils.py:784]:
```suggestion
        def validate_goes(adef, cfinfo):
```

**Comment by djhoese** [comment]
Thanks @zxdawn !
